modules/contact/src/constraints/RANFSNormalMechanicalContact.C

```diff
@@ -0,0 +1,415 @@
+          {
+            auto & master_elem_sequence = _node_to_master_elem_sequence[_current_node->id()];
+            mooseAssert(
+                _current_master == _pinfo->_elem,
+                "The current master element and the PenetrationInfo object's element should "
+                "be the same");
+            master_elem_sequence.push_back(_pinfo->_elem);
+
+            if (master_elem_sequence.size() >= 5 &&
+                _pinfo->_elem == *(master_elem_sequence.rbegin() + 2) &&
+                _pinfo->_elem == *(master_elem_sequence.rbegin() + 4) &&
+                _pinfo->_elem != *(master_elem_sequence.rbegin() + 1) &&
+                *(master_elem_sequence.rbegin() + 1) == *(master_elem_sequence.rbegin() + 3))
+            {
+              // Ok we are ping-ponging. Determine the master node
+              auto master_node =
+                  _penetration_locator._nearest_node.nearestNode(_current_node->id());
+
+              // Sanity checks
+              mooseAssert(_pinfo->_elem->get_node_index(master_node) != libMesh::invalid_uint,
+                          "The master node is not on the current element");
+              mooseAssert((*(master_elem_sequence.rbegin() + 1))->get_node_index(master_node) !=
+                              libMesh::invalid_uint,
+                          "The master node is not on the other ping-ponging element");
+
+              _ping_pong_slave_node_to_master_node.insert(
+                  std::make_pair(_current_node->id(), master_node));
+            }
+          }
+        }
+        else
+          // We have not locked the node into contact nor is the gap smaller than the Lagrange
+          // Multiplier so we should not apply
+          return false;
+      }
+
+      // Determine whether we're going to apply the tied node equality constraint or the contact
+      // inequality constraint
+      auto it = _ping_pong_slave_node_to_master_node.find(_current_node->id());
+      if (it != _ping_pong_slave_node_to_master_node.end())
+      {
+        _tie_nodes = true;
+        _nearest_node = it->second;
+        _master_index = _current_master->get_node_index(_nearest_node);
+        mooseAssert(_master_index != libMesh::invalid_uint,
+                    "nearest node not a node on the current master element");
+      }
+      else
+      {
+        _distance = _pinfo->_distance;
+        // Do this to make sure constraint equation has a positive on the diagonal
+        if (_pinfo->_normal(_component) > 0)
+          _distance *= -1;
+        _tie_nodes = false;
+
+        // The contact constraint is active -> we're going to use our linear solve to ensure that
+        // the gap is driven to zero. We only have one zero-penetration constraint per node, so we
+        // choose to apply the zero penetration constraint only to the displacement component with
+        // the largest magnitude normal
+        auto largest_component_magnitude = std::abs(_pinfo->_normal(0));
```

Comment:
No way to consider all normal components? Should we drive the gap along the normal direction to zero?

Reply:
You only want to introduce one constraint: the gap is zero. By definition, then we can only eliminate one non-linear function. We cannot do such a thing as eliminate n_x of one nonlinear function and n_y of another nonlinear function. We have to make a choice

Reply:
I see